packages/data-point/docs/best-practices.md

```diff
@@ -0,0 +1,457 @@
+
+Patterns related to this topic:
+
+- grouping multiple synchronous reducers together
+- picking/omitting keys from an object
+- using functions to manipulate objects
+- optimizing operations on collections
+
+## Patterns
+
+### Pure Functions
+
+We should think of DataPoint only as a transformation tool, your transformations should(when possible) follow the same principles of [pure functions](https://medium.com/javascript-scene/master-the-javascript-interview-what-is-a-pure-function-d1c076bec976):
+
+- Given the same inputs, always returns the same output.
+- [Referential transparency](https://medium.com/@olxc/referential-transparency-93352c2dd713) - if the expression can be replaced with its value without changing the program’s behavior.
+- Produce no side effects - the expression should not alter any external state (this includes not mutating any of the parameter values).
+
+### Higher Order Reducers
+
+There are multiple types of reducers, you can use them directly on a transformation, or you can use the pattern of [higher order functions](https://medium.com/javascript-scene/higher-order-functions-composing-software-5365cf2cbe99) to create dynamic/parameterized reducers.
+
+- Dynamically create a reducer
+- Parameterize the creation of a reducer
+- Create multiple reducers with similar or default configurations
+
+#### Function reducer
+
+```js
+function getLocale(localeId) {
+  return (input, acc) => {
+    // assume acc.locals.localization holds a Map of all available locales
+    return acc.locals.localization.get(localeId)
+  }
+}
+```
+
+#### Entity Reducer
+
+```js
+const SwapiAPI = (service) => {
+  // generates a Request Entity with name `SwapiAPI-${service}`
+  return Request(`SwapiAPI-${service}`, {
+    // inject the resolved value to the end of the URL
+    url: `https://swapi.co/api/${service}/{value}`,
+    params: {
+      cache: {
+        ttl: '20m'
+        staleWhileRevalidate: '1h'
+      }
+    }
+  })
+}
+
+// for hitting: https://swapi.co/api/people/{value}
+const PeopleRequest = SwapiAPI('people')
+
+// for hitting: https://swapi.co/api/planets/{value}
+const PlanetsRequest = SwapiAPI('planets')
+
+dataPoint.resolve(PeopleRequest, 10).then((result) => {
+  // result would be the value of:
+  // https://swapi.co/api/people/10
+}))
+```
+
+### Picking/omitting keys from an object
+
+The Object reducer can come in very handy when mapping properties from an input. However, its execution is asynchronous where each property it maps is executed asynchronously in parallel; this can potentially be something you might want to avoid to keep things as performant as possible.
```

Comment:
> this can potentially be something you might want to avoid to keep things as performant as possible

this doesn't make it clear *when* you would want to avoid using an object reducer. do you avoid it as much as possible? or only when everything is synchronous?

Reply:
done! thanks!